### ISSUE TITLE
Add full type hover popup to VS Code commands

### ISSUE THREAD

**Issue by Arlen22** [microsoft/TypeScript#35601]
Several issues have been opened for this problem, but they have been closed as a duplicate of an issue which barely mentions the problem, let alone addresses it: https://github.com/microsoft/vscode/issues/64566

The closed issues are:

- https://github.com/microsoft/vscode/issues/66405
- https://github.com/microsoft/vscode/issues/76480
- https://github.com/microsoft/vscode/issues/66405

The problem is simple. The three dots means you can't copy and paste definitions like you used to be able to. This is an absolutely essential feature because TypeScript isn't even close to being used everywhere, and the intellisense of Javascript files inside TypeScript files is way worse than inside Javascript files. This basically means the easiest way to create a d.ts file is to copy the type definition in the hover inside Javascript and then paste it into a d.ts file so TypeScript files can use it. It also helps me debug overloads and unions when things are going awry. 

Even just a command that would popup the hover with the full definition instead of the abbreviated definition would be incredibly useful. Currently the shortcut is `Ctrl+k, Ctrl+i` for the abbreviated definition, so I take it there is a command behind it that can be typed into the command list (`ctrl+shift+p`). So even just adding the command for people to manually add a shortcut for would be really useful. 

I can work around it, but I spend hours more time per week because of it.

**Comment by mjbvz** [comment]
Moving upstream to get more feedback from the TS Team

**Comment by Ciantic** [comment]
This is same request I had in the diagnostics issue.

Use case I had is that I often found myself seeing a type in the hover tip, and copying it out. It never worked after the changes.

**Comment by Arlen22** [comment]
Can we somehow make this a VS Code setting? I guess this is in Typescript now, so whatever the equivalent would be here.

**Comment by Arlen22** [comment]
And now this evening I am diagnosing type problems and it is not showing me a second type that I know is different than the first one that it is showing me. It’s a union on the values of an object and it is only expanding the first type in the union.

**Comment by DanielRosenwasser** [comment]
Just as a heads up, there's a `noErrorTruncation` option which you can set to `true`, which I believe gives you the behavior you're looking for.

**Comment by Ciantic** [comment]
Ah, apparently it doesn't work always.

Btw, this is a duplicate of https://github.com/Microsoft/TypeScript/issues/26238

**Comment by Arlen22** [comment]
I desperately need this feature. Infer types from usage is giving me three dots when I use it if the original type has three dots. I also can't debug type problems because the three dots is not dependable and never guarantees that any particular type is actually repeated. 

Since this is a TypeScript feature I can just build a custom TypeScript package, but I don't know where I would do that. Any ideas?

**Comment by lake2** [comment]
same issue + 1

**Comment by devuxer** [comment]
@mjbvz

> Even in JS there are types that are too large to display (or infinite), so it is not possible to just always expand the ... eagerly.

What about a setting that just allows you to set a max tooltip length, and only apply ellipsis if tooltip length exceeds maximum? This would be a little bit like `Terminal > Integrated: Scrollback`.

**Comment by LoveriusB** [comment]
"noErrorTruncation": true Did not work correctly for me even though it's been proposed several times as a solution in other questions like this one.

A quick fix that actually does work would be... some kind of a more offensive solution ;p 
Start by opening <Microsoft VS Code install folder>/resources/app/extensions/node_modules/typescript/lib/tsserver.js and change ts.defaultMaximumTruncationLength = 160 at around line 13471 (so far) to something higher like ts.defaultMaximumTruncationLength = 800.

This will get you what you are waiting for. A nice 800 chars long max type definition on hover. You can of course custom this value as much as you want.

This might not be politically correct for you will need to do it again after every vs code update, but it works just fine and you'll finally get your full type on hover.
Note that you'll need to restart your tsServer. To do so: 

in vscode: ctrl + shift + P -> type "restart ts server" -> press enter. wait a couple seconds, it's fairly quick.

**Comment by Rechdan** [comment]
As @LoveriusB mentioned above, we can change the value of `defaultMaximumTruncationLength`, but it feels super wrong and a better option, maybe, would be to add that setting to `tsconfig`?

**Comment by tonyxiao** [comment]
Confirmed that this works really great https://github.com/microsoft/TypeScript/issues/35601#issuecomment-921752554

Combined with patch-package and it becomes automatic. Thanks for the awesome workaround until we get this in core.

**Comment by Arlen22** [comment]
Also going back to my OP, isn't there a way that Typescript files could make use of intellisense for Javascript files the same way other Javascript files can? Even just setting the import to `const myJS = require("./myJS.js") as import("./myJS.js")` would be helpful. I'm surprised this is still an issue. Unless I'm missing something.

**Comment by szulcus** [comment]
Any updates?

**Comment by AustinBernerEvercast** [comment]
+1

**Comment by toto1384** [comment]
+1 :))

**Comment by pauldvu** [comment]
Found myself going back and making the edit every so often so I made a bash this bash command to copy and paste into your terminal. Swipe it, it comes in handy!

```bash
:'
Mac OS | Visual Studio Code
Replace 1000 with any number
'
value='defaultMaximumTruncationLength = 1000'
sudo sed -i '' "s/defaultMaximumTruncationLength = 160/$value/" "/Applications/Visual Studio Code.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js"
```
For those of you using Visual Studio Code Insiders
```bash
:'
Mac OS | Visual Studio Code Insiders
Replace 1000 with any number
'
value='defaultMaximumTruncationLength = 1000'
sudo sed -i '' "s/defaultMaximumTruncationLength = 160/$value/" "/Applications/Visual Studio Code - Insiders.app/Contents/Resources/app/extensions/node_modules/typescript/lib/tsserver.js"
```

**Comment by fgarit-te** [comment]
+1 on this being a really useful feature. Right now I'm using the "Typescript Explorer" vscode extension but would hope this to be built in.

**Comment by nikelborm** [comment]
For those plusoners I just want to remind that you by doing that also sending emails to at least 15 people...

**Comment by Raf-A2F** [comment]
- 

- [ ] 
![New Project](https://github.com/microsoft/TypeScript/assets/153484899/3a3b953f-c27a-47bb-ac57-c56c20ff7a9b)
@Arlen22 
`/`
![New Project](https://github.com/microsoft/TypeScript/assets/153484899/c51c5d39-2aa1-47ad-bab9-968b88992fb0)
``

**Comment by shwarcu** [comment]
This is extremely annoying when working with long types that are calculated and cannot be checked in source
<img width="1033" alt="Screenshot 2024-02-05 at 10 25 26" src="https://github.com/microsoft/TypeScript/assets/58426925/2f76915f-98bf-40a2-b78c-6a44b9733cd0">

**Comment by nikelborm** [comment]
For those who haven't seen this plugin:
```
Name: TypeScript Explorer
Id: mxsdev.typescript-explorer
Description: Full type information for variables, components, functions, and more in TypeScript projects!
Version: 0.4.2
Publisher: mxs
```
https://marketplace.visualstudio.com/items?itemName=mxsdev.typescript-explorer

**Comment by wysez** [comment]
> For those plusoners I just want to remind that you by doing that also sending emails to at least 15 people...

Yeah that's the point....

**Comment by heysokam** [comment]
> For those who haven't seen this plugin:
> https://marketplace.visualstudio.com/items?itemName=mxsdev.typescript-explorer

Thank you for sharing. Nextlevel useful :writing_hand:

**Comment by vennson** [comment]
Long term solution by [@extremegf](https://github.com/microsoft/TypeScript/issues/35601#issuecomment-1824306948)

in vscode
1. Using Ctrl+P, Go to `node_modules/typescript/lib/tsserver.js`
2. Find line `var defaultMaximumTruncationLength = 160`
3. Change it to `1000` or value of your choosing.
4. Open Ctrl+Shift+P menu, select TypeScript: Select TypeScript Version...
5. Ensure it's set to Use Workspace Version
6. Full restart VSCode - reloading workspace is not enough.
7. Verify that ... abbreviations are gone from type hints.

PS.
cause `noErrorTruncation` in tsconfig didn't work for me.
and changing tsserver.js of VScode's typescript is anoying every VSCode update

Before:
with ...
<img width="342" alt="image" src="https://github.com/microsoft/vscode/assets/28529262/55a53baf-011e-439d-a358-fa853e0c3e81">


After:
You can also drag the bottom border
<img width="345" alt="image" src="https://github.com/microsoft/vscode/assets/28529262/40cb26bb-d402-46fc-a1d6-dbde143f483c">

**Comment by kimhwanhoon** [comment]
> "noErrorTruncation": true Did not work correctly for me even though it's been proposed several times as a solution in other questions like this one.
> 
> A quick fix that actually does work would be... some kind of a more offensive solution ;p Start by opening /resources/app/extensions/node_modules/typescript/lib/tsserver.js and change ts.defaultMaximumTruncationLength = 160 at around line 13471 (so far) to something higher like ts.defaultMaximumTruncationLength = 800.
> 
> This will get you what you are waiting for. A nice 800 chars long max type definition on hover. You can of course custom this value as much as you want.
> 
> This might not be politically correct for you will need to do it again after every vs code update, but it works just fine and you'll finally get your full type on hover. Note that you'll need to restart your tsServer. To do so:
> 
> in vscode: ctrl + shift + P -> type "restart ts server" -> press enter. wait a couple seconds, it's fairly quick.

I'm useing typescript v5, I found that on `tsc.js` not `tsserver.js`.

**Comment by wertania** [comment]
+1

**Comment by FedericoBiccheddu** [comment]
It used to work in TS 5.4, but now it doesn't (I tried every minor version back to 5.0.4). Even in 5.6, with the following patch, complex types are truncated:
```patch
diff --git a/lib/tsc.js b/lib/tsc.js
index 37e7f28602e6337b6216810d6a40e6701a99f0a5..6ac55bc88de4c3415cac5e218b2b4dad0a3adfc1 100644
--- a/lib/tsc.js
+++ b/lib/tsc.js
@@ -12463,7 +12463,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 1e5;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;
```
Results in:
```
ManagedRuntime<GuestClient | AuthenticatedClient | BaseURL | Session, never>.runPromise: <JWTPayload, _.InvalidJWT>(effect: Effect.Effect<JWTPayload, _.InvalidJWT, GuestClient | AuthenticatedClient | BaseURL | Session>, options?: {
    ...;
} | undefined) => Promise<...>
```

I suspect this is related to VS Code updates, even though I'm using the TypeScript workspace version.

**Comment by SinghChinmay** [comment]
> Long term solution by [@extremegf](https://github.com/microsoft/TypeScript/issues/35601#issuecomment-1824306948)
> 
> in vscode
> 
> 1. Using Ctrl+P, Go to `node_modules/typescript/lib/tsserver.js` or `node_modules/typescript/lib/typescript.js`
> 2. Find line `var defaultMaximumTruncationLength = 160`
> 3. Change it to `1000` or value of your choosing.
> 4. Open Ctrl+Shift+P menu, select TypeScript: Select TypeScript Version...
> 5. Ensure it's set to Use Workspace Version
> 6. Full restart VSCode - reloading workspace is not enough.
> 7. Verify that "..." abbreviations are gone from type hints.
> 
> PS.
> 
> * cause `noErrorTruncation` in tsconfig didn't work for me.
> * changing tsserver.js of VScode's typescript is anoying every VSCode update
> * the extension mentioned in other solution is not copy-pasteable
> 
> Before: with ... <img alt="image" width="342" src="https://private-user-images.githubusercontent.com/28529262/344517259-55a53baf-011e-439d-a358-fa853e0c3e81.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxOTUwMTksIm5iZiI6MTcyNzE5NDcxOSwicGF0aCI6Ii8yODUyOTI2Mi8zNDQ1MTcyNTktNTVhNTNiYWYtMDExZS00MzlkLWEzNTgtZmE4NTNlMGMzZTgxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDE2MTgzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM2ZGYzZTcyZDdiN2UyODJhNzRjYzUxNDNmZGEzY2M2ODUyYzk1ZGZmMTJkY2YyMmE4NTAwZWIwNmMzYzY1MjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AVX-MPPj7fpZO1K_FyIk_C5OI5hxfJpDBITS_nznSfk">
> 
> After: You can also drag the bottom border <img alt="image" width="345" src="https://private-user-images.githubusercontent.com/28529262/344517136-40cb26bb-d402-46fc-a1d6-dbde143f483c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjcxOTUwMTksIm5iZiI6MTcyNzE5NDcxOSwicGF0aCI6Ii8yODUyOTI2Mi8zNDQ1MTcxMzYtNDBjYjI2YmItZDQwMi00NmZjLWExZDYtZGJkZTE0M2Y0ODNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI0VDE2MTgzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg4NWE4ZmRjNDVkMGM1MDZhZjE1NTU3YWUzM2ZiZDYzMDIxYWM0ZWE5NzZiMTA2NmQyYmMzZmYyMWI5NzNlMmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.nPLcRvrzUcS9RahjMl7noxk8nlMraPR0iUfHHqMeUW0">

Thanks, I was halfway making my extension for vs code, you saved me a lot of time.

**Comment by extremegf** [comment]
> Thanks, I was halfway making my extension for vs code, you saved me a lot of time.

I would still happily use your extension :)

**Comment by pauldvu** [comment]
Back to update my workaround for everybody who wants this functionality.

If you have a bun as your package manager you can use the `bun patch` command in order to 
ensure that your modifications  are re-applied after each `bun install`


## Step 1: Run Bun Patch 
```zsh
bun patch typescript@5.8.0-dev.20241113
```

## Step 2:  Edit The Package

./patches/typescript@5.8.0-dev.20241113.patch

```diff
diff --git a/node_modules/typescript/.DS_Store b/.DS_Store
new file mode 100644
index 0000000000000000000000000000000000000000..f9f60d0a8f94f19a9f55d4bcbfeaf7866207ada5
Binary files /dev/null and b/.DS_Store differ
diff --git a/lib/typescript.js b/lib/typescript.js
index 4ed5056a2ff6648b59de5d2a871dab64e8c97a66..ccda9e88105bac6e5003210d98bb4c6cef867aa6 100644
--- a/lib/typescript.js
+++ b/lib/typescript.js
@@ -16116,7 +16116,7 @@ function isInternalDeclaration(node, sourceFile) {
 // src/compiler/utilities.ts
 var resolvingEmptyArray = [];
 var externalHelpersModuleNameText = "tslib";
-var defaultMaximumTruncationLength = 160;
+var defaultMaximumTruncationLength = 30000;
 var noTruncationMaximumTruncationLength = 1e6;
 function getDeclarationOfKind(symbol, kind) {
   const declarations = symbol.declarations;

```
Step 3: Commit The Patch


```zsh
 bun patch --commit 'node_modules/typescript'
```





Troubleshooting 


Prees Command Shift P 

![Image](https://github.com/user-attachments/assets/df42c4df-1db6-4ad9-9644-6eed535347ad)


Search " Typescript Select  "
*If you don't see this option open a typescript file to active typescript in vscode

![Image](https://github.com/user-attachments/assets/73a78f0a-9142-4ced-a640-6c62235d319f)


Lastly  restart the language server

![Image](https://github.com/user-attachments/assets/eb821616-c330-41ec-86ad-20ecfde26a6d)

**Comment by mb21** [comment]
Anybody else where changing `js/ts.hover.maximumLength` in vscode doesn't do anything, or at least not for this case?

<img width="256" height="245" alt="Image" src="https://github.com/user-attachments/assets/af926ec2-3027-4dce-aea2-db52ed5f1cce" />

**Comment by p-bakker** [comment]
Same here, but is due to Astro extension for me, see https://github.com/microsoft/vscode/issues/259321
